### ISSUE TITLE
Conda-installing sbpy causes tests to fail on collection for Python<3.11

### ISSUE THREAD

**Issue by astronomerritt** [dirac-institute/sorcha#353]
**UPDATE 2023-03-22:** The issue below occurs _when sbpy is conda-installed_. Nothing to do with pytables. Setting up the conda environment with every package needed, and then pip-installing sbpy, works fine. Leaving this here so nobody misunderstands.

HDF5 output needs the pytables package to work, but in versions of Python<3.11, installing pytables creates a dependency conflict at worst and at best causes pytest to fail to collect every single test with the following numpy warning:

`RuntimeWarning: numpy.ndarray size changed, may indicate binary incompatibility. Expected 88 from C header, got 96 from PyObject`

I have tried this with Python 3.10, 3.9, 3.8 and 3.7. 3.6 works but this causes other issues. 3.11 so far works fine as long as a conda environment is created following the below instructions.

Sorcha itself runs fine despite this issue, possibly because the warnings are triggering in the background.

### To replicate:

Pull the sorcha repo and cd into it.

Create a new conda environment using:

`conda create -n sspp_test -c conda-forge -c mjuric python=3.10 spiceypy=4.0.1 openorb numpy pandas matplotlib spice-utils pip pytest astropy scipy sbpy pytables`

and activate it.

Then run `pip install -e .'.

Then run pytest from inside the repo folder.

### Notes:
- From the traceback, it looks like the problem is with the astropy.modelling functions called by sbpy. astropy.modeling is calling astropy.table, and therein triggers the warning.

**Comment by astronomerritt** [comment]
`/opt/miniconda3/envs/sspp_test_37/lib/python3.7/importlib/__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
<frozen importlib._bootstrap>:1006: in _gcd_import
    ???
<frozen importlib._bootstrap>:983: in _find_and_load
    ???
<frozen importlib._bootstrap>:953: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:219: in _call_with_frames_removed
    ???
<frozen importlib._bootstrap>:1006: in _gcd_import
    ???
<frozen importlib._bootstrap>:983: in _find_and_load
    ???
<frozen importlib._bootstrap>:953: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:219: in _call_with_frames_removed
    ???
<frozen importlib._bootstrap>:1006: in _gcd_import
    ???
<frozen importlib._bootstrap>:983: in _find_and_load
    ???
<frozen importlib._bootstrap>:953: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:219: in _call_with_frames_removed
    ???
<frozen importlib._bootstrap>:1006: in _gcd_import
    ???
<frozen importlib._bootstrap>:983: in _find_and_load
    ???
<frozen importlib._bootstrap>:967: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:677: in _load_unlocked
    ???
<frozen importlib._bootstrap_external>:728: in exec_module
    ???
<frozen importlib._bootstrap>:219: in _call_with_frames_removed
    ???
surveySimPP/__init__.py:1: in <module>
    from . import modules
surveySimPP/modules/__init__.py:3: in <module>
    from . import PPCalculateApparentMagnitude
surveySimPP/modules/PPCalculateApparentMagnitude.py:3: in <module>
    from .PPCalculateApparentMagnitudeInFilter import PPCalculateApparentMagnitudeInFilter
surveySimPP/modules/PPCalculateApparentMagnitudeInFilter.py:6: in <module>
    from sbpy.photometry import HG, HG1G2, HG12_Pen16, LinearPhaseFunc
/opt/miniconda3/envs/sspp_test_37/lib/python3.7/site-packages/sbpy/photometry/__init__.py:5: in <module>
    from .core import *
/opt/miniconda3/envs/sspp_test_37/lib/python3.7/site-packages/sbpy/photometry/core.py:16: in <module>
    from astropy.modeling import (Fittable1DModel, Parameter)
/opt/miniconda3/envs/sspp_test_37/lib/python3.7/site-packages/astropy/modeling/__init__.py:11: in <module>
    from . import models
/opt/miniconda3/envs/sspp_test_37/lib/python3.7/site-packages/astropy/modeling/models.py:9: in <module>
    from .core import custom_model, hide_inverse, fix_inputs # pylint: disable=W0611
/opt/miniconda3/envs/sspp_test_37/lib/python3.7/site-packages/astropy/modeling/core.py:32: in <module>
    from astropy.table import Table
/opt/miniconda3/envs/sspp_test_37/lib/python3.7/site-packages/astropy/table/__init__.py:4: in <module>
    from .column import Column, MaskedColumn, StringTruncateWarning, ColumnInfo
/opt/miniconda3/envs/sspp_test_37/lib/python3.7/site-packages/astropy/table/column.py:22: in <module>
    from ._column_mixins import _ColumnGetitemShim, _MaskedColumnGetitemShim
astropy/table/_column_mixins.pyx:1: in init astropy.table._column_mixins
    ???
E   RuntimeWarning: numpy.ndarray size changed, may indicate binary incompatibility. Expected 80 from C header, got 88 from PyObject
_________________________________________________ ERROR collecting surveySimPP/modules/tests/test_PPreadColoursUser.py __________________________________________________
/opt/miniconda3/envs/sspp_test_37/lib/python3.7/importlib/__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
<frozen importlib._bootstrap>:1006: in _gcd_import
    ???
<frozen importlib._bootstrap>:983: in _find_and_load
    ???
<frozen importlib._bootstrap>:953: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:219: in _call_with_frames_removed
    ???
<frozen importlib._bootstrap>:1006: in _gcd_import
    ???
<frozen importlib._bootstrap>:983: in _find_and_load
    ???
<frozen importlib._bootstrap>:953: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:219: in _call_with_frames_removed
    ???
<frozen importlib._bootstrap>:1006: in _gcd_import
    ???
<frozen importlib._bootstrap>:983: in _find_and_load
    ???
<frozen importlib._bootstrap>:953: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:219: in _call_with_frames_removed
    ???
<frozen importlib._bootstrap>:1006: in _gcd_import
    ???
<frozen importlib._bootstrap>:983: in _find_and_load
    ???
<frozen importlib._bootstrap>:967: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:677: in _load_unlocked
    ???
<frozen importlib._bootstrap_external>:728: in exec_module
    ???
<frozen importlib._bootstrap>:219: in _call_with_frames_removed
    ???
surveySimPP/__init__.py:1: in <module>
    from . import modules
surveySimPP/modules/__init__.py:3: in <module>
    from . import PPCalculateApparentMagnitude
surveySimPP/modules/PPCalculateApparentMagnitude.py:3: in <module>
    from .PPCalculateApparentMagnitudeInFilter import PPCalculateApparentMagnitudeInFilter
surveySimPP/modules/PPCalculateApparentMagnitudeInFilter.py:6: in <module>
    from sbpy.photometry import HG, HG1G2, HG12_Pen16, LinearPhaseFunc
/opt/miniconda3/envs/sspp_test_37/lib/python3.7/site-packages/sbpy/photometry/__init__.py:5: in <module>
    from .core import *
/opt/miniconda3/envs/sspp_test_37/lib/python3.7/site-packages/sbpy/photometry/core.py:16: in <module>
    from astropy.modeling import (Fittable1DModel, Parameter)
/opt/miniconda3/envs/sspp_test_37/lib/python3.7/site-packages/astropy/modeling/__init__.py:11: in <module>
    from . import models
/opt/miniconda3/envs/sspp_test_37/lib/python3.7/site-packages/astropy/modeling/models.py:9: in <module>
    from .core import custom_model, hide_inverse, fix_inputs # pylint: disable=W0611
/opt/miniconda3/envs/sspp_test_37/lib/python3.7/site-packages/astropy/modeling/core.py:32: in <module>
    from astropy.table import Table
/opt/miniconda3/envs/sspp_test_37/lib/python3.7/site-packages/astropy/table/__init__.py:4: in <module>
    from .column import Column, MaskedColumn, StringTruncateWarning, ColumnInfo
/opt/miniconda3/envs/sspp_test_37/lib/python3.7/site-packages/astropy/table/column.py:22: in <module>
    from ._column_mixins import _ColumnGetitemShim, _MaskedColumnGetitemShim
astropy/table/_column_mixins.pyx:1: in init astropy.table._column_mixins
    ???
E   RuntimeWarning: numpy.ndarray size changed, may indicate binary incompatibility. Expected 80 from C header, got 88 from PyObject`

**Comment by astronomerritt** [comment]
Full traceback:

```python

/_________________________________________________ ERROR collecting surveySimPP/modules/tests/test_PPreadColoursUser.py __________________________________________________
/opt/miniconda3/envs/sspp_test_37/lib/python3.7/importlib/__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
<frozen importlib._bootstrap>:1006: in _gcd_import
    ???
<frozen importlib._bootstrap>:983: in _find_and_load
    ???
<frozen importlib._bootstrap>:953: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:219: in _call_with_frames_removed
    ???
<frozen importlib._bootstrap>:1006: in _gcd_import
    ???
<frozen importlib._bootstrap>:983: in _find_and_load
    ???
<frozen importlib._bootstrap>:953: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:219: in _call_with_frames_removed
    ???
<frozen importlib._bootstrap>:1006: in _gcd_import
    ???
<frozen importlib._bootstrap>:983: in _find_and_load
    ???
<frozen importlib._bootstrap>:953: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:219: in _call_with_frames_removed
    ???
<frozen importlib._bootstrap>:1006: in _gcd_import
    ???
<frozen importlib._bootstrap>:983: in _find_and_load
    ???
<frozen importlib._bootstrap>:967: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:677: in _load_unlocked
    ???
<frozen importlib._bootstrap_external>:728: in exec_module
    ???
<frozen importlib._bootstrap>:219: in _call_with_frames_removed
    ???
surveySimPP/__init__.py:1: in <module>
    from . import modules
surveySimPP/modules/__init__.py:3: in <module>
    from . import PPCalculateApparentMagnitude
surveySimPP/modules/PPCalculateApparentMagnitude.py:3: in <module>
    from .PPCalculateApparentMagnitudeInFilter import PPCalculateApparentMagnitudeInFilter
surveySimPP/modules/PPCalculateApparentMagnitudeInFilter.py:6: in <module>
    from sbpy.photometry import HG, HG1G2, HG12_Pen16, LinearPhaseFunc
/opt/miniconda3/envs/sspp_test_37/lib/python3.7/site-packages/sbpy/photometry/__init__.py:5: in <module>
    from .core import *
/opt/miniconda3/envs/sspp_test_37/lib/python3.7/site-packages/sbpy/photometry/core.py:16: in <module>
    from astropy.modeling import (Fittable1DModel, Parameter)
/opt/miniconda3/envs/sspp_test_37/lib/python3.7/site-packages/astropy/modeling/__init__.py:11: in <module>
    from . import models
/opt/miniconda3/envs/sspp_test_37/lib/python3.7/site-packages/astropy/modeling/models.py:9: in <module>
    from .core import custom_model, hide_inverse, fix_inputs # pylint: disable=W0611
/opt/miniconda3/envs/sspp_test_37/lib/python3.7/site-packages/astropy/modeling/core.py:32: in <module>
    from astropy.table import Table
/opt/miniconda3/envs/sspp_test_37/lib/python3.7/site-packages/astropy/table/__init__.py:4: in <module>
    from .column import Column, MaskedColumn, StringTruncateWarning, ColumnInfo
/opt/miniconda3/envs/sspp_test_37/lib/python3.7/site-packages/astropy/table/column.py:22: in <module>
    from ._column_mixins import _ColumnGetitemShim, _MaskedColumnGetitemShim
astropy/table/_column_mixins.pyx:1: in init astropy.table._column_mixins
    ???
E   RuntimeWarning: numpy.ndarray size changed, may indicate binary incompatibility. Expected 80 from C header, got 88 from PyObject

```

**Comment by astronomerritt** [comment]
Versions (for Python 3.10):

astropy  5.2.1
numpy  1.24.1
sbpy  0.3.0

**Comment by astronomerritt** [comment]
I've nailed this down to one problem: sbpy in the conda install.

If I install all packages with conda apart from sbpy, _then_ pip install sbpy, everything works.

**Comment by astronomerritt** [comment]
Well, I've had a fun time with this. The old error is gone. In its place, running tests creates a new error:

```
../../../../opt/miniconda3/envs/sspp_test_numpy/lib/python3.9/site-packages/asdf/schema.py:14
  /opt/miniconda3/envs/sspp_test_numpy/lib/python3.9/site-packages/asdf/schema.py:14: DeprecationWarning: jsonschema.exceptions.RefResolutionError is deprecated as of version 4.18.0. If you wish to catch potential reference resolution errors, directly catch referencing.exceptions.Unresolvable.
    from jsonschema.exceptions import RefResolutionError, ValidationError
```

**TL;DR:** 

- Only sbpy version 0.3.0 is available on Conda. 
- Conda installing sbpy and spiceypy=4.0.1 in the same environment installs sbpy 0.3.0, which installs asdf version 2.14.3, which leads to the above warning. 
- Conda installing sbpy without spiceypy=4.0.1 installs asdf version 2.15.0, which does not cause the warning. 
- Pip installing sbpy installs sbpy version 0.4.0, which doesn't appear to rely on asdf at all. 
- Once we move away from OIF we won't need spiceypy at all and this warning will go away.

Read on for boring details.

asdf is a package only used by sbpy in our build. This message appears when asdf version 2.14.3 is installed and goes away if asdf version 2.15.0 is installed. However, sbpy isn't the problem here. The problem is spiceypy. If you create a conda environment with sbpy and without spiceypy=4.0.1: 

`conda create -n conda_test_1 -c conda-forge -c mjuric python=3.9 openorb numpy pandas matplotlib spice-utils pip pytest sbpy pytables setuptools=66.0.0`

conda installs **asdf version 2.15.0**, and the tests run. 

If you create a conda environment with both sbpy and spiceypy=4.0.1:

`conda create -n conda_test_2 -c conda-forge -c mjuric python=3.9 openorb numpy pandas matplotlib spice-utils pip pytest sbpy pytables setuptools=66.0.0 spiceypy=4.0.1`

conda installs **asdf version 2.14.3** and the tests produce the above warning. Now spiceypy doesn't depend on asdf itself, so I can only assume there's a shared dependency there messing everything up.

If you try to conda install asdf=2.15.0, this forces an upgrade of numpy to 1.25.1, which kills difi, because difi uses Numba and that needs numpy<1.24.

So what does this have to do with pip? Well, **pip installing sbpy doesn't install asdf at all**, as far as I can tell, which obviously circumvents the problem completely. However, pip-installing sbpy installs sbpy version 0.4.0 while conda is still installing 0.3.0, so maybe it's no longer a requirement. I can't tell from the release notes but it seems from a quick look at the repo that asdf isn't required for the most recent version.

I don't know if moving to the latest version of spiceypy would resolve this. However, if we are dumping OIF, then this will actually cease to be an issue since we won't need spiceypy any more, and we can merrily conda install sbpy without issues.

I have only tested this with Python 3.9, for the record.

**Comment by mschwamb** [comment]
spicypy is something needed for OIF. So I think we can separate this out and say if you're going to run OIF you should use a separate conda environment. So I think that then solves this issue. Is that correct @astronomerritt?

**Comment by astronomerritt** [comment]
That would solve the problem pretty neatly. (Especially because I've since discovered that another OIF dependency, setuptools=66.0.0, causes the same issue.)

I'd recommend then something like this for Sorcha:

`conda create -n sorcha -c conda-forge python=3.9 numpy pandas matplotlib pip sbpy pytables scipy astropy`

That covers all the main Sorcha dependencies... except for difi, which isn't available via Conda. (The repo lists a conda install, but it's from source after cloning the repo.)

**Comment by astronomerritt** [comment]
Disregard the above, the difi repo's README is out-of-date and difi _is_ published to conda. Testing it.

**Comment by astronomerritt** [comment]
Disregard the above, I am a jetlagged idiot. Difi is published to conda. Testing it now.

**Comment by astronomerritt** [comment]
Okay, that works. A conda environment for Sorcha can be created that contains _all_ of the main dependencies:

`conda create -n sorcha -c conda-forge -c moeyensj python=3.9 numpy pandas matplotlib pip sbpy pytables scipy astropy difi`

Running `pip install -e .` thus doesn't pip install any packages. Tests run, Sorcha runs, all is good.

**Comment by mschwamb** [comment]
I've conda installed with pytables,  sorcha,  sbpy, and spicyepy with no problem. So In think we're over this issue as we move away from OIF as a default. If someone wants to use OIF they can do so separately in a different conda environment to running sorcha
